src/metabase/api/api_key.clj

```diff
@@ -29,28 +30,83 @@
+                                                          :name         name
+                                                          :unhashed_key unhashed-key
+                                                          :created_by   api/*current-user-id*})
+                          (t2/hydrate :group_name))]
+          (events/publish-event! :event/api-key-create
+                                 {:object  api-key
+                                  :user-id api/*current-user-id*})
+          (-> api-key
+              (select-keys [:created_at :updated_at :id :group_name :name])
+              (assoc :unmasked_key unhashed-key
+                     :masked_key (api-key/mask unhashed-key))))))))
 
 (api/defendpoint GET "/count"
   "Get the count of API keys in the DB"
   [:as _body]
   (api/check-superuser)
   (t2/count :model/ApiKey))
 
+(api/defendpoint PUT "/:id"
+  "Update an API key by changing either its group or its name"
+  [id :as {{:keys [group_id name] :as _body} :body}]
+  {id       ms/PositiveInt
+   group_id [:maybe ms/PositiveInt]
+   name     [:maybe ms/NonBlankString]}
+  (api/check-superuser)
+  (let [api-key-before (-> (t2/select-one :model/ApiKey :id id)
+                           ;; hydrate the group_name for audit logging
+                           (t2/hydrate :group_name))]
+    (when name
+      (t2/with-transaction [_conn]
+        ;; A bit of a pain to keep these in sync, but oh well.
+        (t2/update! :model/User (:user_id api-key-before) {:first_name name})
+        (t2/update! :model/ApiKey id {:name name})))
+    (when group_id
+      (let [user (-> api-key-before (t2/hydrate :user) :user)]
+        (user/set-permissions-groups! user [(perms-group/all-users) {:id group_id}])))
+    (let [updated-api-key (-> (t2/select-one :model/ApiKey :id id)
+                              (t2/hydrate :group_name))]
+      (events/publish-event! :event/api-key-update {:object updated-api-key
+                                                    :previous-object api-key-before
+                                                    :user-id api/*current-user-id*})
+      (select-keys updated-api-key [:created_at :updated_at :id :name :masked_key :group_name]))))
+
+(api/defendpoint PUT "/:id/regenerate"
+  "Regenerate an API Key"
+  [id]
+  {id ms/PositiveInt}
+  (api/check-superuser)
+  (let [api-key-before (t2/select-one :model/ApiKey id)
+        unhashed-key (key-with-unique-prefix)
+        _ (t2/update! :model/ApiKey :id id {:unhashed_key unhashed-key})
+        api-key (-> (t2/select-one :model/ApiKey id)
+                    (t2/hydrate :group_name))]
+    (events/publish-event! :event/api-key-regenerate
+                           {:object api-key
+                            :previous-object api-key-before
+                            :user-id api/*current-user-id*})
+    (-> (t2/select-one :model/ApiKey id)
+        (t2/hydrate :group_name)
```

Comment:
This seems like it's doing more DB calls than it needs to. Why do you need to re-select the ApiKey and hydrate it again after publishing the `api-key-regenerate` event?

Reply:
You probably don't even need to re-select it at all tbh. Just do this
* Read the API key
* Update the `:unhashed_key` on the object and write it back
* Use the updated object to generate the API response without needing to read it again

It would be nice if there was a `t2/update-returning-instances` but looks like that [hasn't been implemented yet](https://github.com/camsaul/toucan2/blob/c5346586c85f8dcd25573696d3e2f976b6ac0550/src/toucan2/update.clj#L65)

Reply:
Nice, thanks for noticing this.